### PR TITLE
Implement `kani::Arbitrary` for `Box<T>`

### LINKED ISSUE #2403: Feature request: Implement `kani::Arbitrary` for `Box<T>`
Requested feature: Implement `kani::Arbitrary` for `Box<T>`
Use case: I'm experimenting with user code that defines a struct similar to:
```rust
struct Foo<T> {
    boxed: Box<T>,
}
```
If I attempt to `#[derive(kani::Arbitrary)]` on that struct, I get this error:

```
error[E0277]: the trait bound `std::boxed::Box<T>: kani::Arbitrary` is not satisfied
   --> test.rs:9:5
    |
9   |     boxed: Box<T>,
    |     ^^^^^^^^^^^^^ the trait `kani::Arbitrary` is not implemented for `std::boxed::Box<T>`
    |
    = help: the following other types implement trait `kani::Arbitrary`:
              ()
              (A, B)
              (A, B, C)
              (A, B, C, D)
              (A, B, C, D, E)
              (A, B, C, D, E, F)
              (A, B, C, D, E, F, G)
              (A, B, C, D, E, F, G, H)
            and 39 others
note: required by a bound in `kani::any`
   --> /home/ubuntu/kani/library/kani/src/lib.rs:114:15
    |
114 | pub fn any<T: Arbitrary>() -> T {
    |               ^^^^^^^^^ required by this bound in `any`

error: aborting due to previous error
```

Since `Box<T>` comes from Rust, the Kani library should implement `Arbitrary` for it (similar to #2224 / #2225).

### PR DESCRIPTION
**Pull request by adpaco-aws**
### Description of changes: 

This PR implements `Arbitrary` for `Box<T>`. This is required for structs that `#[derive(kani::Arbitrary)]` and have members with `Box<T>` types.

### Resolved issues:

Resolves #2403 

### Call-outs:

 - I haven't thought too much about the best way to test these changes, so I'm open to suggestions.
 - The same-purpose tests added in #2225 were in the `ui` folder. I moved them to `expected` because that seems to be a better place (no UI testing is done in those tests).

### Testing:

* How is this change tested? Existing tests + new one.

* Is this a refactor change? No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/kani/src/arbitrary.rs [hunk 1]
```diff
@@ -161,3 +161,12 @@ impl Arbitrary for std::marker::PhantomPinned {
         PhantomPinned
     }
 }
+
+impl<T> Arbitrary for std::boxed::Box<T>
+where
+    T: Arbitrary,
+{
+    fn any() -> Self {
+        Box::new(T::any())
```
Comment (zhassan-aws) [inline library/kani/src/arbitrary.rs:170]: Does this cover all possible states a `Box` can have? What about a box created using `Box::new_uninit`?

File: tests/expected/derive-arbitrary/box/test.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+#[derive(kani::Arbitrary)]
+struct Foo<T> {
+    boxed: Box<T>,
+}
+
+#[kani::proof]
+fn main() {
+    let foo: Foo<i32> = kani::any();
+    kani::cover!(*foo.boxed >= i32::MIN && *foo.boxed <= i32::MAX);
```
Comment (zhassan-aws) [inline tests/expected/derive-arbitrary/box/test.rs:15]: It might be better to cover specific corner values, e.g. `kani::cover!(*foo.boxed == i32::MIN)`. Same for 0 and `i32::MAX`.

**Review by zhassan-aws — COMMENTED** [review]
I'm concerned about the soundness of the `Arbitrary` implementation. We need to make sure the implementation covers all possible states a box can be in. I'm a bit unclear on whether this is the case with the current implementation.

File: library/kani/src/arbitrary.rs (hunk 1 above)
Comment (celinval) [reply library/kani/src/arbitrary.rs:170]: `Box::new_uninit` wouldn't fit this signature though, since T would have to be `MaybeUninit<U>`

**Review by celinval — APPROVED** [review]

File: tests/expected/derive-arbitrary/box/test.rs (hunk 2 above)
Comment (adpaco-aws) [reply tests/expected/derive-arbitrary/box/test.rs]: Done: split this cover statement into the three cases you mentioned.
